NeoML/test/src/ReferenceDnnTest.cpp

```diff
@@ -0,0 +1,232 @@
+/* Copyright @ 2024 ABBYY
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+--------------------------------------------------------------------------------------------------------------*/
+
+#include <common.h>
+#include <mutex>
+#pragma hdrstop
+
+#include <TestFixture.h>
+
+using namespace NeoML;
+using namespace NeoMLTest;
+
+namespace NeoMLTest {
+
+struct CReferenceDnnTestParam {
+	CDnn* net;
+};
+
+static void runDnn(int, void* params)
+{
+	CReferenceDnnTestParam* taskParams = static_cast<CReferenceDnnTestParam*>(params);
+	taskParams->net->RunOnce();
+}
+
+static CDnn* createDnn(CRandom& random)
+{
+	CDnn* net = new CDnn(random, MathEngine());
+
+	CBaseLayer* layer = Source(*net, "in");
+	layer = FullyConnected(50)("fc1", layer);
+	layer = Dropout(0.1f)("dp1", layer);
+	layer = FullyConnected(200)("fc2", layer);
+	layer = Dropout(0.1f)("dp2", layer);
+	layer = FullyConnected(10)("fc3", layer);
+	layer = Sink(layer, "sink");
+
+	return net;
+}
+
+static void initializeBlob(CDnnBlob* blob, CRandom& random, int min, int max) {
+	for(int j = 0; j < blob->GetDataSize(); ++j) {
+		blob->GetData().SetValueAt(j, static_cast<float>(random.Uniform(min, max)));
+	}
+}
+
+static void getTestDnns(CArray<CDnn*>& dnns, CArray<CRandom>& randoms,
+	bool useReference, const int& numOfThreads)
+{
+	CObjectArray<CDnnBlob> blobs;
+
+	for(int i = 0; i < numOfThreads; ++i) {
+		bool ifCreateNewDnn = (i == 0 || !useReference);
+		if(ifCreateNewDnn) {
+			dnns.Add(createDnn(randoms[i]));
+		} else {
+			dnns.Add(dnns[i-1]->CreateReferenceDnn());
+		}
+
+		CRandom randomInit(0);
+		CPtr<CDnnBlob> srcBlob = CDnnBlob::CreateTensor(MathEngine(), CT_Float, { 1, 1, 1, 8, 20, 30, 10 });
+		initializeBlob(srcBlob.Ptr(), randomInit, 0, 1);
+		static_cast<CSourceLayer*>(dnns[i]->GetLayer("in").Ptr())->SetBlob(srcBlob);
+	}
+}
+
+static void runMultithreadInference(CArray<CDnn*>& dnns, const int numOfThreads)
+{
+	CArray<CReferenceDnnTestParam> taskParams;
+	for (int i = 0; i < numOfThreads; ++i) {
+		taskParams.Add({ dnns[i] });
+	}
+
+	IThreadPool* pool = CreateThreadPool(numOfThreads);
+	for (int i = 0; i < numOfThreads; ++i) {
+		pool->AddTask(i, runDnn, &(taskParams[i]));
+	}
+
+	pool->WaitAllTask();
+	delete pool;
+}
+
+static void perfomanceTest(bool useReference, const int numOfThreads=4)
+{
+	CArray<CDnn*> dnns;
+	CArray<CRandom> randoms;
+
+	for(int i = 0; i < numOfThreads; ++i) {
+		randoms.Add(CRandom(i));
```

Comment:
Are your test works correct? Here are set different seeds for randoms, then in CreateReferenceDnn().
Maybe, you should copy original random in the line 66 instead of creation here.